### PR TITLE
[PLAT-218][Needs WB PR] Use WB download action callbacks to tally preprint views

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

We want to use addons.base.views.create_waterbutler_log to tally preprint views, this allows us to do that.

**Remember this is for views, not downloads, and only views from non-contributors.**

## Changes

- remove update_analitics from osfstorage_download function
- Change log behavior to increment download_counts during waterbutler callback.
- Add tests.

## QA Notes

You will have to run the migrations and check via the shell as there's nothing user facing in this PR. Also remember contributor downloads don't count in the total.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/PLAT-218
Ticket with more discussion info:
https://openscience.atlassian.net/browse/PLAT-388
Waterbutler PR:
https://github.com/CenterForOpenScience/waterbutler/pull/324

### CODE REVIEW

File: addons/base/views.py [hunk 1]
```diff
@@ -313,14 +314,25 @@ def create_waterbutler_log(payload, **kwargs):
     with transaction.atomic():
         try:
             auth = payload['auth']
-            # Don't log download actions
-            if payload['action'] in ('download_file', 'download_zip'):
+            user = OSFUser.load(auth['id'])
+            # Don't log download actions,
+            if payload['action'] == 'download_zip':
```
Comment (sloria) [inline addons/base/views.py]: Drive-by: This can go above the OSFUser query (to prevent an unused query).

File: osf/models/preprint_service.py [hunk 2]
```diff
@@ -41,6 +41,7 @@ class PreprintService(DirtyFieldsMixin, GuidMixin, IdentifierMixin, ReviewableMi
 
     identifiers = GenericRelation(Identifier, related_query_name='preprintservices')
     preprint_doi_created = NonNaiveDateTimeField(default=None, null=True, blank=True)
+    download_count = models.IntegerField(default=0)
```
Comment (sloria) [inline osf/models/preprint_service.py]: Drive-by: Shouldn't we use the existing `PageCounter` model for this?

Comment (Johnetordoff) [reply osf/models/preprint_service.py]: @sloria It's a bullet point in the original ticket that this field be created, I'm unsure exactly why.

Comment (sloria) [reply osf/models/preprint_service.py]: Ah, those were the notes from grooming. We just forgot about the `PageCounter` class.

File: addons/base/views.py [hunk 3]
```diff
@@ -313,14 +313,20 @@ def create_waterbutler_log(payload, **kwargs):
     with transaction.atomic():
         try:
             auth = payload['auth']
+            user = OSFUser.load(auth['id'])
             # Don't log download actions
             if payload['action'] in ('download_file', 'download_zip'):
+                if payload['action_meta']['is_mfr_render']:
+                    node = AbstractNode.load(payload['metadata']['nid'])
+                    if not node.is_contributor(user):
+                        file = BaseFileNode.load(payload['metadata']['path'].lstrip('/'))
```
Comment (erinspace) [inline addons/base/views.py]: np: file is a builtin name in python, so not a good variable name choice, but shouldn't actually be hurting anything

File: osf/models/files.py [hunk 4]
```diff
@@ -108,6 +108,8 @@ class BaseFileNode(TypedModel, CommentableMixin, OptionalGuidMixin, Taggable, Ob
     objects = BaseFileNodeManager()
     active = ActiveFileNodeManager()
 
+    mfr_views = models.PositiveIntegerField(default=0)
```
Comment (erinspace) [inline osf/models/files.py]: np: perhaps better organized above the manager declarations and with the other fields, so they're not obscured, but nothing technically wrong with this

Comment (erinspace) [reply osf/models/files.py]: Is there a reason you didn't use the `PageCounter` class mentioned in the ticket? I saw there was a discussion in a previous comment but wasn't sure of the resolution as it still seems to be in the ticket.

Comment (Johnetordoff) [reply osf/models/files.py]: I'd be open to using the PageCounter, but this just seems more intuitive to me. I don't know what the pros of using the PageCounter would be. I think if someone is looking for the view count they'd expect to find it on the file model.

Comment (sloria) [reply osf/models/files.py]: The problem with incrementing a field on the file is that the record gets locked every time the file gets viewed, increasing the probabiility of a deadlock. Using `PageCounter` solves this problem and also saves a timestamp for views, which is is important for internal metrics.

File: osf/models/files.py [hunk 5]
```diff
@@ -328,6 +327,18 @@ def get_download_count(self, version=None):
 
         return count or 0
 
+    def get_view_count(self, version=None):
+        """Pull the mfr view count from the pagecounter collection
+        Limit to version if specified.
+        """
+        parts = ['view', self.node._id, self._id]
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/files.py]: 1. 👍 
2. There's no user facing or API facing changes to this.

File: addons/base/views.py [hunk 6]
```diff
@@ -313,14 +314,24 @@ def create_waterbutler_log(payload, **kwargs):
     with transaction.atomic():
         try:
             auth = payload['auth']
-            # Don't log download actions
+            # Don't log download actions, but do update analytics
+            user = OSFUser.load(auth['id'])
             if payload['action'] in ('download_file', 'download_zip'):
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply addons/base/views.py]: Yeah, it's confusing downloads here just means downloaded from waterbutler, it doesn't make a distinction whether it was an MFR view etc. This whole section of the code was originally only adding to the logs, so the info isn't very nuanced in the way we'd like it to be for analytics.

File: addons/base/views.py [hunk 7]
```diff
@@ -311,14 +312,24 @@ def create_waterbutler_log(payload, **kwargs):
...
+                node = AbstractNode.load(payload['metadata']['nid'])
+                if not node.is_contributor(user):
+                    url = furl.furl(payload['request_metadata']['request']['url'])
```
Comment (felliott) [inline addons/base/views.py]: The WB-side PR needed for this has been merged, but one change will be needed.  The request url is now in the `payload['request_meta']['url']` key (along with `referrer`, `user_agent`, and `method`).

**Comment by sloria** [comment]
@Johnetordoff Can you look into the failing tests? They are the ones that were added in this PR.

Once the failing tests are addressed this should be good to merge.

**Comment by sloria** [comment]
Merged this into the wrong branch 2 days ago, so I reverted this and then reverted the revert on next-next-platform. tl;dr This is on `release/next-next-platform` now.
